### ISSUE TITLE
Updating `multidms` fitting

### ISSUE THREAD

**Issue by jgallowa07** [dms-vep/dms-vep-pipeline-3#86]
`multidms==3.0` introduces a much easier interface for fitting collections of models in parallel. @Caleb-Carr and I have put together this issue to serve as an outline/plan for updating the notebooks using `multidms`. 

## How it's being used

Currently, there are three notebooks which use the multidms package:

1. [global epistasis](https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/notebooks/func_effects_global_epistasis.ipynb) fit only to a reference. i.e. no shifts
2. [individual replicates shifts](https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/notebooks/func_effect_shifts.ipynb) fitting individual replicate groups of experimental conditions to compare.
3. [averaging and plotting](https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/notebooks/avg_func_effect_shifts.ipynb) notebook which takes the shift parameters produced in the previous notebook and aggregates them for visualization with the polyclonal.lineplot_and_heatmap utility.

The parameters which these notebooks use to fit can be found in the [config file](https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/test_example/data/func_effects_config.yml). Note that there are two config files - the `multidms` relevant parameters are in the data/func_effects.yaml, generally.

To see the output of the notebook runs on some real data, the Lassa virus analysis page can be found [here](https://dms-vep.github.io/LASV_Josiah_GP_DMS/). The notebook runs are in the "functional effects of mutations" -> "analysis notebooks" drop downs.

For development, we will be using the test example as described in the [README](https://github.com/dms-vep/dms-vep-pipeline-3#overview).

## What should change

###. Using `fit_one_model` interface, and specifying softplus output clipping.

So here ([global epistasis](https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/notebooks/func_effects_global_epistasis.ipynb) ), the model is being fit on a single experiment at a time, boiling it down to a simple global epistasis model - great. However, it looks like each of the fits has some form of functional score clipping of the input functional scores before fitting, along with a default model which only has a linear (identity) output activation. To make use of the model prediction clipping feature, you must instantiate the model with the "Softplus" output activation function, which itself is parameterized by the "lower_bound" parameter of the `Model` class.
This problem persists in the [individual replicates shifts](https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/notebooks/func_effect_shifts.ipynb) notebook.

So, in summary, you have:
```
# initialize with default params, which give sigmoid global epistasis function
model = multidms.Model(data)
model.fit()
```
but probably want to use the new `model_collection.fit_one_model` interface which is slightly simpler, and more interpret-able.
```python
model_fit_params = {
    "dataset": data,
    "output_activation" : "Softplus",
    "lower_bound" : global_epistasis_params[f"clip_{bound}"]
}
multidms.fit_one_model(**model_fit_params)
```
This will produce a `pandas.Series` like
```
model                              Model(Model-0)
dataset_name                          Replicate 1
step_loss                    [0.6847866920000956]
epistatic_model                           Sigmoid
output_activation                        Softplus
scale_coeff_lasso_shift                   0.00001
scale_coeff_ridge_beta                          0
scale_coeff_ridge_shift                         0
scale_coeff_ridge_gamma                         0
scale_coeff_ridge_alpha_d                   0.001
huber_scale_huber                               1
gamma_corrected                             False
alpha_d                                      True
init_beta_naught                              0.0
lock_beta_naught_at                          None
tol                                        0.0001
num_training_steps                              1
iterations_per_step                         20000
n_hidden_units                                  5
lower_bound                                  <X>
PRNGKey                                         0
dtype: object
```
From here you can access the `multidms.Model` object for all the plotting utilities etc.

The rest of the notebook looks good, but we will have to update the name of the predicted_func_score columns produced by `Model.get_mutations_df()` from `f"{selection}_predicted_func_score": "functional_effect",` -> `f"predicted_func_score_{selection}": "functional_effect",`

### Using `ModelCollection` for lasso sweep

Everything up to the model initialization and fitting cell look good. For fitting, we'll now want to use the more convenient [multidms.fit_models](https://matsengrp.github.io/multidms/multidms.html#multidms.fit_models) utility. Note that we also want to be doing the Softplus activation with the clipped predictions.

```python
lasso_shifts = params["lasso_shifts"]
assert len(lasso_shifts) == len(set(lasso_shifts))
n_lasso = len(params["lasso_shifts"])
fig, ax = plt.subplots(n_lasso, 2, figsize=[6, n_lasso * 3])

model = multidms.Model(data)
mutations_df = []
for i, lasso_shift in enumerate(lasso_shifts):
    lasso_shift = float(lasso_shift)
    print(f"Fitting model for {lasso_shift=}")
    model.fit(lasso_shift=lasso_shift)
    mutations_df.append(
        model.get_mutations_df(phenotype_as_effect=True).assign(lasso_shift=lasso_shift)
    )
    model.plot_epistasis(ax=ax[i, 1], alpha=0.1, show=False, legend=not i)
    model.plot_pred_accuracy(ax=ax[i, 0], alpha=0.1, show=False, legend=False)
    ax[i, 1].set_title(f"Epistasis fit (lasso {lasso_shift})")
    ax[i, 0].set_title(f"Accuracy (lasso {lasso_shift})")

plt.show()
```
Will get changed to something like
```python
lasso_shifts = params["lasso_shifts"]
assert len(lasso_shifts) == len(set(lasso_shifts))
n_lasso = len(params["lasso_shifts"])
fig, ax = plt.subplots(n_lasso, 2, figsize=[6, n_lasso * 3])

model_collection_params = {
    "dataset": [data],
    "output_activation" : ["Softplus"]  # could also include, "Identity" in this list if you wanted to test diffent combinations of models,
    "lower_bound" : [global_epistasis_params[f"clip_{bound}"]]
    "scale_coeff_lasso_shift" : params["lasso_shifts"]
}
fits = multidms.fit_models(**model_collection_params)
```
This will result in a dataframe which essentially stacks the Series objects as seen above. So for creating the epistasis plots, we'll simply loop through the rows of the dataframe and call the plotting method on the `Model` attribute of the row. You could also plot the histogram of shift parameters here.

### `ModelCollection` sparsity visualization.

Once you have a dataframe containing the individual fits across lasso weights and whatever else, you can initialize the `ModelCollection` object and use the methods for visualizing the results of model sparsity as a function of lasso weight. Examples of these plots can be seen in the `multidms` [documentation](https://matsengrp.github.io/multidms/fit_delta_BA1_example.html#ModelCollection-class). For example, given the example above, we could do:
```python
mc = multidms.ModelCollection(fits)
sparsity_chart = mc.shift_sparsity(times_seen_threshold=3)
sparsity_chart
```
to get something like 
![Screenshot from 2023-11-03 12-40-00](https://github.com/dms-vep/dms-vep-pipeline-3/assets/23087427/5b7767e9-00d9-4c39-baf6-89daf9b1b37d)

Note: we should make sure that the relevant plotting parameters are accessible via the config file, and have reasonable defaults.

The final step of the second notebook should be a `pickle` dumped binary representation of the `fits` DataFrame (not the `ModelCollection` object, for reasons explained below). 

 ### Final aggregation of replicates.

Each of the individual replicate notebooks that get run will produce a DataFrame object. The [final notebook](https://github.com/dms-vep/dms-vep-pipeline-3/blob/main/notebooks/avg_func_effect_shifts.ipynb) will then collect each of the dataframes, stack (horizontal concatination) these dataframes into a single dataframe with all replicate fits and lasso weights combined. Something like

```python
replicate_fits_df = pd.concat(
   [pickle.load(f"replicate_fits_{x}.pkl")
   for x in params["comparisons"]
)
mc = multidms.ModelCollection(replicate_fits_df)
```
From here, you can make the same sparsity plots as shown above (but now you are averaging between the different replicates).
```python
sparsity_chart = mc.shift_sparsity(times_seen_threshold=3)
sparsity_chart
```

Additionally, as well as the averaged heatmap for a specified lasso strengths (note it doesn't make sense to average accoss fits with different lasso stregths)
```python
chart = mc.mut_param_heatmap(
    times_seen_threshold=3,
    query="scale_coeff_lasso_shift == 5e-5",
    mut_param="shift" # or "shift"
)
chart
```
Here, we could loop and make a chart for every lasso strength if desired, as well as making a separate heatmap for each parameter ("beta", "shift", "predicted_func_score").

### Misc / ?:
* We need to make sure to assign a unique name to each of the `Data` Objects as they're getting initialized. This will persist through the collections so that we may easily compare and aggregate fits downstream.
* Make sure that any parameters, or configurations changed line up with the respective (https://github.com/dms-vep/dms-vep-pipeline-3/blob/a3125b98223b66dadfeace83505cb2daff0d2d2e/func_effects.smk#L303)[snakemake rules].

**Comment by jbloom** [comment]
@jgallowa07, great, thanks for posting this.

I suggest we split the updates related to fitting the individual global epistasis models and the shifts into two separate pull requests, unless that will create problems with versioning of `multidms` in the `conda` environment (can we use the new global epistasis models for one model while still using old code for shifts if we do this two step?).

For the single global epistasis model, I'm actually not 100% clear why we need the softplus output clipping if we are just clipping the input values at a reasonable lower bound. Does it not work to just have those input values be the "real data" and not invoke softplus output clipping? I know you have thought about this a lot more than me so I'm probably wrong in my thinking here, but at least wanted to ask question.

**Comment by jbloom** [comment]
Sounds good, yes---I do think a small pull request to do the softplus clipping is great. If you can update `multidms` to use the `fit_one_model` interface without breaking the multi-condition `multidms`, that might even be way to go there.

Then maybe a separate pull request for the multi-condition?

**Comment by jgallowa07** [comment]
@jbloom sorry for the late follow-up. 

> If you can update multidms to use the fit_one_model interface without breaking the multi-condition multidms, that might even be way to go there.

Unfortunately, the interface was released in [version 3.0.0](https://github.com/matsengrp/multidms/blob/main/CHANGELOG.rst) along with some breaking-naming changes, and thus would require updating the multi-condition notebooks. That said, I have tested the softplus prediction clipping on the /test_example - The code diff is quite trivial as pointed out above. Similarly trivial differences between resulting fits as well. 

w/o lower bound predcition clipping (main/HEAD)

![Screenshot from 2023-11-20 15-18-11](https://github.com/dms-vep/dms-vep-pipeline-3/assets/23087427/76e9dba4-2be4-4fcb-b70f-ccc7e3c24e18)

w/ softplus clipping

![Screenshot from 2023-11-20 15-22-19](https://github.com/dms-vep/dms-vep-pipeline-3/assets/23087427/1983691d-a142-4fc4-9f7f-fa06f9fff136)





I'm happy to submit a PR. Alternatively, I could lump this all together (since the GE code will be overwritten in the next PR, as well) and be sure to expose the relevant options for model behavior in the config file. I can validate that the defaults for these things remain at the current state the pipeline.  

Either way, looks like I would need to fork the repo with my current privileges - which is totally ok if that's the route you prefer

**Comment by jbloom** [comment]
Great, I just gave you write access to the repo.

I'm happy to have you do either 1 or 2 pull requests as you think is best.

But for the global epistasis fitting, could I ask one favor first? Would you also be able to re-run the original and new model (similar to above) on a "real" dataset, such as @Caleb-Carr's LASV or @Bernadetadad's XBB.1.5 spike to confirm similar results too? The test example is a very small dataset so may not be overly realistic of real ones. If you don't want to re-run their entire pipelines, they can also probably help you find files just for notebooks of interest. Or you should just be able to clone and re-run entire repos provided you have access to read the FASTQs of the `bloom_j_grp` on the Hutch server (I'm not sure of this).

**Comment by jgallowa07** [comment]
Great - I just pushed to the branch [86_1_multidms_softplus_clipping](https://github.com/dms-vep/dms-vep-pipeline-3/tree/86_1_multidms_softplus_clipping). 

Sure, @Caleb-Carr would you be willing to help test this out? you could either rebase your submodule to test this out with the branch above, or, simply provide me with one or more of the functional scores csv's that would be found in results/func_scores/*_func_scores.csv.

**Comment by jgallowa07** [comment]
So just as a quick peak at running a single replicate (LibA-220823-293T-1) [Caleb's 293T Data](https://dms-vep.github.io/LASV_Josiah_GP_DMS/notebooks/func_effects_global_epistasis_LibA-220823-293T-1.html) - we can see an example of the sigmoid floor dropping below the lower bound of the clipped data.

w/o lower bound prediction clipping (from the linked results above)

![Screenshot from 2023-11-21 14-50-25](https://github.com/dms-vep/dms-vep-pipeline-3/assets/23087427/025591e5-48b2-4240-890f-5e74419b56fa)

w/ softplus clipping

![Screenshot from 2023-11-21 14-30-09](https://github.com/dms-vep/dms-vep-pipeline-3/assets/23087427/09a300c2-9617-412e-b214-c29107b8f171)

I can't say without further exploration that this is an empirically "better" fit, but it is quite similar. Additionally, using the softplus prediction clipping resulted in a the functional score distribution for mutations to non-stop amino acids seem more sensible. e.g.

w/o lower bound prediction clipping (from the linked results above)

![Screenshot from 2023-11-21 14-37-30](https://github.com/dms-vep/dms-vep-pipeline-3/assets/23087427/8472dd8e-5917-4423-af9d-815401080563)

Here, I notice a "shelf", or bimodality to the the blue distribution. Whereas with softplus clipping, the distribution is probably more what we would expect

![Screenshot from 2023-11-21 14-48-18](https://github.com/dms-vep/dms-vep-pipeline-3/assets/23087427/ca9d3561-43b5-4806-8fc4-c22f31428946)


Happy to run the rest of the replicates, just not sure what granularity/format of the results you'd most prefer if so.

**Comment by jbloom** [comment]
These are probably similar enough we could just change. Although I do notice for the softplus clipipng, more of the orange (stop) distribution moves to the right around or greater than zero, which is not a good sign probably. But this effect is small, so not sure if it is meaningful. Any thoughts on what to do next for this?

**Comment by jgallowa07** [comment]
Good catch, yea that is a little disconcerting. Because further model validation is in the works anyway (re manuscript revisions), it's likely we'll have better ways to determine and evaluate the model type you would like to run in this pipeline. I suggest that we get the interfaces updated for all three multidms notebooks (following the basic outline up top) and then circle back to deciding the best default model choice?

As mentioned, this update can expose parameters (e.g. like "softplus_pred_clipping") and set their defaults to whatever you prefer. I'm agnostic to softplus prediction clipping, it's just what we found helpful with the Spike data IIRC.

**Comment by jbloom** [comment]
For the new interface, are the results identical to the old ones without softplus clipping?

I wonder if we could set up initially with the new interface but no clipping for sort of backward compatibility, and then add an option to add the clipping and people can evaluate that for their datasets and maybe eventually we change default?

**Comment by jgallowa07** [comment]
> For the new interface, are the results identical to the old ones without softplus clipping?

Indeed, it is!

> initially with the new interface but no clipping for sort of backward compatibility

Yup - agreed! I'm on it.
